sqlsmith: enhance tlp to test predicate projections and DISTINCTs

jordanlewis opened the pull request:
This commit enhances the WHERE tlp generator to also output the
projection of the predicates in both the partitioned and unpartitioned
query outputs. The purpose of this enhanced testing is to make sure that
projection output always matches filter output, and that projection
output for all 3 partitions is identical in every partition and in the
non-partitioned case as well. 

Also, add a DISTINCT oracle.

The DISTINCT oracle for TLP generates unpartitioned queries of the
following form:

SELECT DISTINCT {cols...} FROM t

and partitioned queries of the following form:

SELECT DISTINCT {cols...} FROM t WHERE pred UNION ALL
SELECT DISTINCT {cols...} FROM t WHERE NOT (pred) UNION ALL
SELECT DISTINCT {cols...} FROM t WHERE (pred) IS NULL

Release note: None

jordanlewis committed:
sqlsmith: enhance tlp to test predicate projections

This commit enhances the WHERE tlp generator to also output the
projection of the predicates in both the partitioned and unpartitioned
query outputs. The purpose of this enhanced testing is to make sure that
projection output always matches filter output, and that projection
output for all 3 partitions is identical in every partition and in the
non-partitioned case as well.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73701)
<!-- Reviewable:end -->

jordanlewis commented:
Hi @mgartner, I wrote this extra testing as a way to start understanding TLP a bit better. I think it does test some new things, since the path for projections is different from the path for filters. But, maybe it's not novel enough to add? I'm not sure, interested to see what you think.

jordanlewis committed:
sqlsmith: implement DISTINCT TLP oracle

The DISTINCT oracle for TLP generates unpartitioned queries of the
following form:

SELECT DISTINCT {cols...} FROM t

and partitioned queries of the following form:

SELECT DISTINCT {cols...} FROM t WHERE pred UNION ALL
SELECT DISTINCT {cols...} FROM t WHERE NOT (pred) UNION ALL
SELECT DISTINCT {cols...} FROM t WHERE (pred) IS NULL

Release note: None

jordanlewis commented:
Second commit adds the `DISTINCT` oracle given by Manuel's paper.

jordanlewis commented:
Gentle ping on this if you have some time to take a look!

mgartner reviewed (approved):
Very cool! Sorry for the delay. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> , just left some minor suggestions.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-:-Mv-hEXo7yEDeeE91qX4:bdotadk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/internal/sqlsmith/tlp.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-fHZx02Nkl5GNoqIj:-Mv-fHZx02Nkl5GNoqIk:bdxb8ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad834146a1640e5d120c44b509f15857da25d540/pkg/internal/sqlsmith/tlp.go#L65)):*
> ```Go
> // The first query returned is an unpartitioned query of the form:
> //
> //   SELECT *, p, NOT (p), (p) IS NULL, true, false, false FROM table
> ```

I'm not sure the last three columns are necessary. Is there a type of bug you have in mind that they would catch that would not be caught by the addition of the first three columns?
___
*[pkg/internal/sqlsmith/tlp.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-fd1o8_VYWZ8t7pTT:-Mv-fd1o8_VYWZ8t7pTU:bme3chr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad834146a1640e5d120c44b509f15857da25d540/pkg/internal/sqlsmith/tlp.go#L109)):*
> ```Go
> FROM %s
> WHERE %s`, allPreds, pred1, pred2, pred3, tableName, pred1)
> 	part2 := fmt.Sprintf(`SELECT *, 
> ```

nit: remove trailing whitespace
___
*[pkg/internal/sqlsmith/tlp.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-gbPW3QaqUwXo8rsP:-Mv-gbPX2Xw-WfjLuchl:b-arkx2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d37605320030fa4ac8b1a62bf8797e2f6201bc9/pkg/internal/sqlsmith/tlp.go#L392)):*
> ```Go
> //
> //   SELECT DISTINCT {cols...} FROM table WHERE (p) UNION ALL
> //   SELECT DISTINCT {cols...} FROM table WHERE NOT (p) UNION ALL
> ```

If you change the `UNION ALL`s to `UNION`s, then you don't need the to return a `distinct` boolean above and you can remove de-duplicating code from `unsortedMatrixesDiff`.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73701)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/internal/sqlsmith/tlp.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-fHZx02Nkl5GNoqIj:-Mv0CO1MDwtmXUl26TvX:b5s082d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad834146a1640e5d120c44b509f15857da25d540/pkg/internal/sqlsmith/tlp.go#L65)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not sure the last three columns are necessary. Is there a type of bug you have in mind that they would catch that would not be caught by the addition of the first three columns?
</blockquote></details>

I added this because if there's some way an entire column was be projected to the incorrect boolean, we'd have false passes. I guess it's an implausible case though?
___
*[pkg/internal/sqlsmith/tlp.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-gbPW3QaqUwXo8rsP:-Mv0CKLwDLZjMDZCf0t_:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d37605320030fa4ac8b1a62bf8797e2f6201bc9/pkg/internal/sqlsmith/tlp.go#L392)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

If you change the `UNION ALL`s to `UNION`s, then you don't need the to return a `distinct` boolean above and you can remove de-duplicating code from `unsortedMatrixesDiff`.
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-:-Mv0wFVTAJACMLKgsQx-:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/internal/sqlsmith/tlp.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-fHZx02Nkl5GNoqIj:-Mv0vfZFCJ6u9Wn5S3iW:b-thjnk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad834146a1640e5d120c44b509f15857da25d540/pkg/internal/sqlsmith/tlp.go#L65)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I added this because if there's some way an entire column was be projected to the incorrect boolean, we'd have false passes. I guess it's an implausible case though?
</blockquote></details>

:shrug: Hard to say how plausible that is. I'm fine keeping it. Might be explaining the columns in this comment though.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-:-Mv1eBR86NThPef-IiZ9:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/internal/sqlsmith/tlp.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73701#-Mv-fHZx02Nkl5GNoqIj:-Mv1eA5FATH8TRUKPjQJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad834146a1640e5d120c44b509f15857da25d540/pkg/internal/sqlsmith/tlp.go#L65)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

:shrug: Hard to say how plausible that is. I'm fine keeping it. Might be explaining the columns in this comment though.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+
